### PR TITLE
RBD Regional DR: Implement clusterID mapping 

### LINKED ISSUE #2118: Add clusterID and poolID configuration to support failover in Regional DR
```0001-0009-rook-ceph-0000000000000002-b0285c97-a0ce-11eb-8c66-0242ac110002```

The Above is the volumeID is sent back in the response to the CreateVolume operation and will be added as a volume handle in the PV spec. CO (Kubernetes) uses above as the identifier for further operations on the volume/PVC. 

The VolumeID is encoded as below

```
0001 -->	                         [csi_id_version=1:4byte] + [-:1byte]
0009 -->	                          [length of clusterID=1:4byte] + [-:1byte]
rook-ceph -->	                          [clusterID:36bytes (MAX)] + [-:1byte]
0000000000000002 -->	                  [poolID:16bytes] + [-:1byte]
b0285c97-a0ce-11eb-8c66-0242ac110002 --> [ObjectUUID:36bytes]

Total of constant field lengths, including '-' field separators would hence be,
4+1+4+1+1+16+1+36 = 64
```

When mirroring is enabled the volume which is `csi-vol-ObjectUUID` is mirrored to the other cluster.

> `csi-vol` is const name and over has the option to override it in storageclass.

 During the Disaster Recovery (failover operation) the PVC and PV will be recreated on the other cluster when cephcsi receives the request for operations like (NodeStage, ExpandVolume, DeleteVolume, etc) the volumeId is sent in the request which will help to identify the volume.

```yaml=
apiVersion: v1
kind: ConfigMap
data:
  config.json: |-
    [
      {
       "clusterID": "rook-ceph",
        "radosNamespace": "<rados-namespace>",
        "monitors": [
          "192.168.39.82:6789"
        ],
        "cephFS": {
          "subvolumeGroup": "<subvolumegroup for cephfs volumes>"
        }
      },
            {
       "clusterID": "fs-id",
        "radosNamespace": "<rados-namespace>",
        "monitors": [
          "192.168.39.83:6789"
        ],
        "cephFS": {
          "subvolumeGroup": "<subvolumegroup for cephfs volumes>"
        }
      }
    ]
metadata:
  name: ceph-csi-config
```

During CSI operations, CephCSI will decode the volumeID and gets the monitor configuration from the configmap, and from the pool ID it will get the pool Name and retrieves the OMAP data stored in the rados OMAP, and finally checks the volume is present in that pool.

We have two below problems here

* The clusterID can be different
    * as the clusterID is the namespace where rook is deployed, the Rook might be deployed in a different namespace on the secondary cluster
    * In standalone cephcsi the clusterID is fs ID. and fs ID is unique per cluster

* The poolID can be different
    * Pools ID which is encoded in the volumeID won't remain the same across clusters

### PR DESCRIPTION
**Pull request by Madhu-1**
Whenever Ceph-CSI receives a CSI/Replication request it will first decode the volumeHandle and try to get the required OMAP details if it is not able to retrieve receives a `Not Found` error message and Ceph-CSI will check for the clusterID mapping. If the old volumeID `0001-00013-site1-storage-0000000000000001-b0285c97-a0ce-11eb-8c66-0242ac110002`
contains the `site1-storage` as the clusterID, now Ceph-CSI will look for the corresponding clusterID `site2-storage` from the clusterID mapping file. If the clusterID mapping is found now Ceph-CSI will look for the poolID mapping ie mapping between `1` and `2`. Example:- pool with name exists on both the clusters with different IDs Replicapool with ID `1` on site1 and Replicapool
with ID `2` on site2.  After getting the required mapping Ceph-CSI has the required information to get more details from the rados OMAP. If we have multiple clusterID mapping it will loop through all the mapping and checks the corresponding pool to get the OMAP data. If the clusterID mapping does not exist Ceph-CSI will return a `Not Found` error message to the caller.

Fixes: #2118
Depends-on: #2292

## Note to Reviewers

Up for initial review

## TODO:

- [x] Add example configmap yaml
- [x] Update required deployment templates to mount the configmap as volume
- [x] E2E testing

### CODE REVIEW

File: internal/rbd/rbd_util.go
Comment (nixpanic) [inline internal/rbd/rbd_util.go]: This should probably be in a helper function.

File: internal/util/cluster_mapping.go
Comment (nixpanic) [inline internal/util/cluster_mapping.go:40]: This is under discussion in #2292, maybe mark this PR as a draft?

I have added `Depends-on:` in the PR description, Mergify will now only merge this PR after the other is merged (new feature).

File: internal/util/cluster_mapping.go
Comment (nixpanic) [inline internal/util/cluster_mapping.go:28]: make this a const?

File: internal/util/cluster_mapping.go
Comment (nixpanic) [inline internal/util/cluster_mapping.go]: does this really need to be a configurable path?

File: internal/util/cluster_mapping.go
Comment (nixpanic) [inline internal/util/cluster_mapping.go]: `#nosec` is only needed it the path is non-const, consider making it a const.

File: internal/rbd/rbd_util.go
Comment (nixpanic) [inline internal/rbd/rbd_util.go:1056]: are these PV annotations documented somewhere? is it because the mapping might not be sufficient?

File: internal/rbd/rbd_util.go
Comment (nixpanic) [inline internal/rbd/rbd_util.go:1148]: A comment would help explaining what this functions purpose is

File: examples/csi-config-map-sample.yaml
Comment (nixpanic) [inline examples/csi-config-map-sample.yaml]: indention looks a little off

**Review by nixpanic — CHANGES_REQUESTED** [review]
just a few nits, no real issues as far as I can spot

File: internal/util/cluster_mapping.go
Comment (Madhu-1) [reply internal/util/cluster_mapping.go:28]: added this as var to modify this on in go tests and verify different combinations. if we make it const the value cannot be changed and even we don't have access to create files in the `/etc` folder in go tests.

File: internal/util/cluster_mapping.go
Comment (Madhu-1) [reply internal/util/cluster_mapping.go]: followed the pattern same in csiconfig.go. this can be changed also

File: internal/rbd/rbd_util.go
Comment (Madhu-1) [reply internal/rbd/rbd_util.go:1056]: No, it's not documented. This was the earlier logic to identify the mapping which is added by the controller we start in the provisioner pod.  with new changes mapping is sufficient. Will remove logic of adding/retrieving PV annotation to identify the mapping in next release. will open an issue for the same.

File: examples/csi-config-map-sample.yaml
Comment (Madhu-1) [reply examples/csi-config-map-sample.yaml]: Yes, Thanks for catching this one. Fixed now.

File: internal/util/cluster_mapping.go
Comment (Madhu-1) [reply internal/util/cluster_mapping.go]: updated now not to take this as input param

File: internal/rbd/rbd_util.go
Comment (nixpanic) [reply internal/rbd/rbd_util.go:1056]: @Madhu-1 don't forget to open an issue about this!

**Review by nixpanic — DISMISSED** [review]

File: internal/rbd/rbd_util.go
Comment (Madhu-1) [reply internal/rbd/rbd_util.go:1056]: https://github.com/ceph/ceph-csi/issues/2328 here it is

**Comment by Madhu-1** [comment]
> /retest ci/centos/upgrade-tests-rbd

`X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
`

**Comment by Madhu-1** [comment]
@nixpanic manually rebased the PR and we store FsID not pool ID for cephfs volumes. Updated the examples now PTAL.

**Comment by Madhu-1** [comment]
> /test ci/centos/upgrade-tests-cephfs

`X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute
`

**Comment by Madhu-1** [comment]
@humblec @nixpanic PTAL Thanks.

**Review by nixpanic — DISMISSED** [review]

**Comment by Madhu-1** [comment]
@ShyamsundarR @agarwal-mudit can you please help with a review here?

**Comment by Madhu-1** [comment]
changed FsID to FscID for cephfs. @Rakshith-R Thanks for the suggestion. @Rakshith-R @nixpanic PTAL

**Review by nixpanic — APPROVED** [review]

**Review by Rakshith-R — APPROVED** [review]
Thanks ! Look good to me  :+1: 
(simple + easy to config + highly scale-able :100:  :wink: )

File: internal/rbd/rbd_util.go
Comment (ShyamsundarR) [inline internal/rbd/rbd_util.go:1047]: (nit) should we return err as above or the new mErr as determined here?

**Review by ShyamsundarR — APPROVED** [review]
One small nit, should not affect functionality though.

File: internal/rbd/rbd_util.go
Comment (Madhu-1) [reply internal/rbd/rbd_util.go:1047]: returning `mErr` to propagate the error from GetClusterMappingInfo to the GRPC so that if there are any errors user/admin can fix them.

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by Madhu-1** [comment]
@mergifyio backport release-v3.4

**Pull request merged**
